src/crypto/CHIPCryptoPAL.h

```diff
@@ -1804,6 +1805,65 @@ CHIP_ERROR ExtractSubjectFromX509Cert(const ByteSpan & certificate, MutableByteS
  **/
 CHIP_ERROR ExtractIssuerFromX509Cert(const ByteSpan & certificate, MutableByteSpan & issuer);
 
+class PemEncoder
+{
+public:
+    static constexpr size_t kNumBytesPerLine   = 48u;
+    static constexpr size_t kMinLineBufferSize = 64u + 1u; // PEM expects 64 characters wide and a null terminator at least.
+    static_assert(kMinLineBufferSize == (BASE64_ENCODED_LEN(kNumBytesPerLine) + 1),
+                  "Internal incoherence of library configuration!");
+
+    /**
+     * @brief Construct a PEM encoder for element type `encodedElement` whose DER data is in `derBytes` span.
+     *
+     * @param encodedElement - Element type string to include in header/footer (e.g. "CERTIFICATE"). Caller must provide correct
+     * uppercase.
+     * @param derBytes - Byte span containing data to encode. May be empty.
+     */
+    explicit PemEncoder(const char * encodedElement, ByteSpan derBytes) : mEncodedElement(encodedElement), mDerBytes(derBytes) {}
+
+    // No copies.
+    PemEncoder(const PemEncoder &)             = delete;
+    PemEncoder & operator=(const PemEncoder &) = delete;
+
+    /**
+     * @brief Returns the next line of the encoding, if anything left to do.
+     *
+     * If `destSize` == 0, nothing will be written. Otherwise if `destSize < kMinLineBufferSize`,
+     * only a null-terminator is written. Otherwise the line is filled with the next line
+     * of PEM encoding. Line is guaranteed null-terminated. Even if nothing good
+     * is written, encoder proceeds to the next line.
+     * When header/footer are written, the heading type (`encodedElement` value) such as
+     * `CERTIFICATE` is clamped so that the entire header line with `-----BEGIN ${encodedElement}-----`
+     * and footer line with `-----END ${encodedElement}-----` are not wider than 64 bytes. This
+     * will not happen in practice with the types of things this is meant to encode.
+     *
+     * @param destStr - null-terminated char buffer to receive the line.
+     * @param destSize - size of the char buffer, including null-terminator.
+     * @return true if there is a line given back to accumulate.
+     * @return false if the encoding is complete.
+     */
+    bool NextLine(char * destStr, size_t destSize);
+
+    bool IsDone() const { return mState == State::kDone; }
+
+private:
+    enum State : int
+    {
+        kPrintHeader = 0,
+        kPrintBody   = 1,
+        kPrintFooter = 2,
+        kEndIterator = 3,
+        kDone        = 4,
```

Comment:
It's not really clear how kEndIterator and kDone differ.  Could use some documentation...